### DISCUSSION TITLE
Having 2 or more HelmRelease in the same namespace with the same releaseName will cause race deletion of resources

### DISCUSSION THREAD

**Discussion by cr1cr1** [fluxcd/flux2#2407] [Q&A] [answered]
```yaml
name: HelmRealease1
spec:
    releaseName: myrelease
```
```yaml
name: HelmRealease2
spec:
    releaseName: myrelease
```

Seems to cause looped reconciliation where each one will trigger the deletion of the other's resources, because they do not belong to it.

I assumed `releaseName` was basically just a helm name prefix, but it seems to be more than that.

Is this expected? Of course, I could let `releaseName` blank to use the default but ugly IMHO prefix or use myrelease1 and myrelease2 for each HelmRelease. But then again what is the point of having a unique HelmRelease name and unique releaseName? I did notice the secrets `sh.helm.release.v1.myrelease1.v1` and `sh.helm.release.v1.myrelease2.v1` if using different releaseName.

Did not find anything related to this in the docs, nor issues or here in the discussions (did not perform hours of search though, just tried to search by keywords like releaseName, race, loop, etc).

**Comment by kingdonb** [comment, marked as answer]
The default setting for `HelmRelease.spec.releaseName` is described in the API doc:

`releaseNamestring | (Optional)ReleaseName used for the Helm release. Defaults to a composition of ‘[TargetNamespace-]Name’.`

The main use of `releaseName` is for setting the HelmRelease name equal to an existing HelmRelease that pre-dates your installing Flux v2, so it can be adopted. If you run `helm ls` and look at the names of the releases there, that is what you're setting with `releaseName`.

It might not be clear from the docs. But with that information in mind, it should be clear why you have a problem when you set `releaseName` to the same value on more than one HelmRelease. (Because they are trying to install into the same target release.)

Only one can win. So this might be a bad behavior given the configuration is bad, and perhaps we should fix it so the worst case scenario that I think you're describing doesn't happen, but the main point to get across is this is not a good config and you should not set `releaseName` for two releases in the same `targetNamespace` to the same value, as it will never have a useful result.

Here is the doc that I pulled this information from: https://fluxcd.io/docs/components/helm/api/

The use of structuring releaseName default the way that it is, I think, is that from Flux's perspective, there is little to no risk of accidental name conflicts where two HelmReleases are in the same namespace and they target different `targetNamespace` but do not set `storageNamespace` equal to `targetNamespace`. But what you've done here is intentionally set them to have a conflict (whether you intended to or not, anyway.)

Backing all the way up... generally when I see new Helm Controller users struggling, I recommend them to set a few things on every HelmRelease:

* `spec.timeout` - set this and you can control the timeout, you may want it to be a shorter time than `spec.interval`
* `releaseName` - if you want the Helm release installed with a specific name, set this. I usually set it equal to `metadata.name`
* `targetNamespace` - for historical reasons or centralization reasons, you might keep all Flux resources in `flux-system`. If the HelmRelease should be installed in a different namespace than where the HelmRelease CR lives, set it here.
* `storageNamespace` - to avoid confusion, set this equal to `targetNamespace`.

The `storageNamespace` is the namespace where Helm keeps its storage `secret`. (If this is not equal to `targetNamespace`, then `helm ls -n targetNamespace` and `helm ls -n storageNamespace` will both provide unexpected results, although this is a valid configuration, it's not straightforward to explain why you would want to do this.)

But try it, and take a look at what these commands do to understand the role of these settings better. You can put the Helm secret in a different namespace; this might be useful for some security applications where you don't want the target namespace to have all the information that you gave Helm at install time, (maybe there are some secrets that are privileged and kept in a different namespace.)

I'm not sure if this is well covered already in the HelmRelease guide. Maybe we should add something about it. What docs were you referring to, (so we can get a sense of where you might have most easily found this information if we made sure to add it there?)

**Comment by cr1cr1** [comment]
Thank you for your answer, @kingdonb . I am following the same docs as you have pointed out. Generally, they are really helpful. Where they fall short, I usually look in the source code to get a better idea.

I think this is relevant and it is how I understood it:

> It might not be clear from the docs. But with that information in mind, it should be clear why you have a problem when you set releaseName to the same value on more than one HelmRelease. (Because they are trying to install into the same target release.)

Yes, I have set them intentionally to be the same `releaseName`, assuming it was **just a prefix for resulting objects of a HelmRelease**, but it obviously is used as key metadata to identify a targeted release.

Naively, I was looking at this from a general deployment perspective, aiming to have a HelmRelease for an app and another HelmRelease for a database, both in the same namespace, both resulting in objects prefixed with something like `foo-`. Eventually I have set `releaseName` to `fooapp` and `foodb`  for each one, of course.

The reason I did not let the default `releaseName` that indeed would be safer in avoiding conflicts, is because the resulted object names can be ridiculous, for example, taking the case of `prometheus-stack`, I would end up with something like: `alertmanager-MONITORING-KUBE-PROMETHEUS-STACK-kube-prometheus-stack-alertmanager-0`, where `MONITORING` is a namespace and `KUBE-PROMETHEUS-STACK` is the HelmRelease `medatada.name`. The chart also composes names from metadata files, including the release prefix. Of course, all this fuss is for... aesthetics, and IMHO easier for a human to see objects at a glance (using an UI like Lens or K9S).

What I was also expecting, in the case of a `releaseName` conflict, was some kind of first-come, first-served behavior: First HelmRelease would have the release created, all others resulting in an error because of "release already exists but is not the intended one" because of different metadata than the first one.

**Reply by kingdonb** [reply]
That's not all that `releaseName` is used for, I get that ostensibly the release name is used by the templates to make sure the rendered template of two Helm releases in the same namespace with different names don't overlap, but I think that it's also used for the name of the secret that helm itself keeps release accounting. Helm Controller is actually using Helm under the hood, so everything Helm does it also does.

```
sh.helm.release.v1.openvpn.v1     helm.sh/release.v1                    1      24d
sh.helm.release.v1.openvpn.v2     helm.sh/release.v1                    1      22d
sh.helm.release.v1.openvpn.v3     helm.sh/release.v1                    1      11m
sh.helm.release.v1.openvpn2.v1    helm.sh/release.v1                    1      69s
```

I tried creating a duplicate of my "openvpn" releasename here and simply changing it to releaseName of `openvpn2` in the same namespace, which is the storageNamespace (example linked)

https://github.com/kingdonb/bootstrap-repo/commit/56ce0c1f61a0bb468dc6aed5b2f4cc22f7093baf

You can see this is the original HelmRelease and that is the new one:

```yaml
spec:
  interval: 9m0s
  timeout: 8m0s
  releaseName: openvpn
  targetNamespace: openvpn
  storageNamespace: openvpn
```

```yaml
spec:
  interval: 9m0s
  timeout: 8m0s
  releaseName: openvpn2
  targetNamespace: openvpn
  storageNamespace: openvpn
```

I think you might have better results with what you're trying to do if you set `storageNamespace` to something else. That's not a great solution but it's a workaround that I think you'll be able to live with.

Since Helm does its release accounting in the secret, two helmreleases in the same namespace can't have the same `releaseName` because they will overwrite and conflict each others' storage. It's not a race, each new release actually checks the inventory from the last release for things it doesn't see in a current version of the template, thus should delete, and all the resources being in the other chart turn out to be not found in this version of the template, so Helm itself is garbage collecting them.

But there's no reason it would do that if it wasn't configured with the same `storageNamespace` in both conflicting helmreleases.

So I'm saying you can actually do what you're trying to do, just set one or the other HelmRelease to a different `storageNamespace`. Generally `storageNamespace` and `targetNamespace` should match, but in your case that's why you have an issue I think.

**Comment by cr1cr1** [comment]
Indeed, what you described is the observed behavior. I do not want to achieve what I imagined initially at all costs. I think that a better way would be to have each `releaseName` unique and keep `targetNamespace` the same as `storageNamespace` (the default).

Thank you!
